src/framework/components/rigid-body/system.js

```diff
@@ -239,10 +239,14 @@ Object.assign(pc, function () {
         onRemove: function (entity, data) {
             if (data.body) {
                 this.removeBody(data.body);
+
                 Ammo.destroy(data.body);
-            }
 
-            data.body = null;
+                var motionState = data.body.getMotionState();
+                delete Ammo.btMotionState.__cache__[motionState.ptr];
```

Comment:
This looks odd. Can you add code comments explaining the requirement for the above 2 lines please?

Reply:
When btRigidBody is created, it creates btMotionState for it. And when deleting btRigidBody, it does not deletes its motion state. It is weird, does it reuses them internally - I'm not sure. But deleting it using `Ammo.destroy` will lead to errors. After further testing, just removing reference from cache - does the trick.
I did not went too far, profiling WebAssembly memory, I would assume they still have some objects references stored inside, unfortunately.

After a lot of testing, this has not brought any issue, and cleared JS references to it.

Reply:
Is it worth opening an issue on the Ammo repo and seeing if Alon Zakai has any thoughts?

Reply:
Removed the delete bit, and added an issue on ammo.js